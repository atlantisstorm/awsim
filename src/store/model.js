import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const handleFetchData = ({ state, accessKeyId, secretAccessKey, region }) => {
  let instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D
  if (process.env.NODE_ENV === "development") {
    instances = extractInstances(testInstances);
  } else {
    AWS.config.update({
      "accessKeyId": accessKeyId,
      "secretAccessKey": secretAccessKey,
      "region": region
    });

    const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});
    const params = {
      DryRun: false
    };

    ec2.describeInstances(params, function(err, data) {
      if(err) {
        console.log("Error", err.stack);
      } else {
        instances = extractInstances(data);
      }
    });
  }

  return { ...state, loading: false, instances };
}

export const handleFilterInstances = ({ state }) => {
  const filteredInstances = filterInstances(state.instances, state.filters);
  return { ...state, filteredInstances };
}

export const handleFilterUpdate = ({ state, filterName, filterValue }) => {
  const filters = state.filters;
  const updatedFilter = { ...filters[filterName], currentValue: filterValue };
  const updatedFilters = { ...filters, [filterName]: updatedFilter };
  const filteredInstances = filterInstances(state.instances, updatedFilters);

  return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };    
}

export const handleModalSetDisplay = ({ state, display }) => {
  const newModalSettings = {
    ...state.modalSettings,
    display: display
  };

  return { ...state, modalSettings: newModalSettings };
}

export const handleModalShowInstance = ({ state, instanceId, pageX, pageY}) => {
  const newModalSettings = {
    ...state.modalSettings,
    pageX,
    pageY,
    display: true,    
    instance: getInstanceRawData(state.instances, instanceId)
  };

  return { ...state, selectedInstance: instanceId, modalSettings: newModalSettings };
}

export const handleRefreshData = ({ state }) => {
  return { ...state, loading: true };
}

export const handleSortOptionsSet = ({ state, field, order }) => {
  return { ...state, sortOptions: setSort(field, order) };
}

export const Model = {
  handleFetchData,
  handleFilterInstances,
  handleFilterUpdate,
  handleModalSetDisplay,
  handleModalShowInstance,
  handleRefreshData,
  handleSortOptionsSet
};

export default Model;