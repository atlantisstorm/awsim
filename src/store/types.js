// Emmm... action types.
export const FETCH_DATA = "FETCH_DATA";
export const FILTER_INSTANCES = "FILTER_INSTANCES";
export const FILTER_UPDATE = "FILTER_UPDATE";
export const MODAL_SET_DISPLAY = "MODAL_SET_DISPLAY";
export const MODAL_SHOW_INSTANCE = "MODAL_SHOW_INSTANCE";
export const REFRESH_DATA = "REFRESH_DATA";
export const SORT_OPTIONS_SET = "SORT_OPTIONS_SET";