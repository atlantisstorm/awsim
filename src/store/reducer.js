import { useReducer } from 'react';
import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
import {
  FETCH_DATA,
  FILTER_INSTANCES,
  FILTER_UPDATE,
  MODAL_SET_DISPLAY,
  MODAL_SHOW_INSTANCE,
  REFRESH_DATA,
  SORT_OPTIONS_SET
} from './types';

const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const reducer = (state, action) => {
  let payload;
  let instances;
  let filters;
  let newModalSettings;

  switch (action.type) {
    case FETCH_DATA: {
      payload = action.payload;
      instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D

      if (process.env.NODE_ENV === "development") {
        instances = extractInstances(testInstances);
      } else {
        AWS.config.update({
          "accessKeyId": payload.accessKeyId,
          "secretAccessKey": payload.secretAccessKey,
          "region": payload.region
        });

        const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});

        const params = {
          DryRun: false
        };

        ec2.describeInstances(params, function(err, data) {
          if(err) {
            console.log("Error", err.stack);
          } else {
            instances = extractInstances(data);
          }
        });
      }

      return { ...state, loading: false, instances };
    }

    case FILTER_INSTANCES: {
      const filteredInstances = filterInstances(state.instances, state.filters);
      return { ...state, filteredInstances };
    }

    case FILTER_UPDATE: {
      payload = action.payload;
      filters = state.filters;

      const filterName = payload.name;
      const filterValue = payload.value;
      const updatedFilter = { ...filters[filterName], currentValue: filterValue };
      const updatedFilters = { ...filters, [filterName]: updatedFilter };

      const filteredInstances = filterInstances(state.instances, updatedFilters);
      return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };
    }

    case MODAL_SET_DISPLAY: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: payload.display
      };
      return { ...state, modalSettings: newModalSettings };
    }

    case MODAL_SHOW_INSTANCE: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: true,
        pageX: payload.pageX, 
        pageY: payload.pageY,
        instance: getInstanceRawData(state.instances, payload.instanceId)
      };
      return { ...state, selectedInstance: payload.instanceId, modalSettings: newModalSettings };
    }

    case REFRESH_DATA: {
      // This will trigger FETCH_DATA
      return { ...state, loading: true };
    }

    case SORT_OPTIONS_SET: {
      // This will trigger FETCH_DATA
      payload = action.payload;
      return { ...state, sortOptions: setSort(payload.field, payload.order) };
    }

    default: {
      return state;
    }
  }
};

export const initialiseReducer = (props) => {
  const iState = (props && props.initialState) ? props.initialState : defaultState;
  const [state, dispatch] = useReducer(reducer, iState);
  return { state, dispatch };
}