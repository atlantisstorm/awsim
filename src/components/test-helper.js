import React from 'react';
//import { render } from "@testing-library/react";
import Context from "../store/context";
import { initialiseReducer } from "../store/reducer";

export const RenderWithProvider = (props) => {
  const { children, testState, testDispatch }  = props;
  // Note: we pass any supplied testState to initialiseReducer rather than switching it like
  // we do for dispatch because of scoping.  Basically each rerender after a click/select
  // update event would mean we refresh again with the original data.
  const { state, dispatch: iDispatch } = initialiseReducer({ initialState: testState });
  const dispatch = testDispatch ? testDispatch : iDispatch;
  return (
    <Context.Provider value={{ state, dispatch }}>
      { children }
    </Context.Provider> 
  )
}