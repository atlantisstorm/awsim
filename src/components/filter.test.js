import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import Filter from './filter';
import extractInstances from '../fixtures/extracted-instances';
import { defaultState } from '../store/reducer';
import { FILTER_UPDATE } from '../store/types';
import { RenderWithProvider } from './test-helper';

describe('Filter component', () => {
  const testState = {
    ...defaultState,
    loading: false,
    instances: extractInstances,
    filteredInstances: extractInstances
  };

  it('should match snapshot', () => {
    const expectedFilter = {
      name: 'Name',
      field: 'Name',
      type: 'static',
      match: 'regex',
      values: [
        'ftp',
        'image',
        'loghost',
        'minion',
        'rx',
        'sql',
        'tx',
        'workhorse'
      ],
      includeEmptyDefaultValue: true
    };

    const filter = testState.filtersOrdered[0];
    expect(filter).toMatchObject(expectedFilter);

    const { asFragment } = render(
      <RenderWithProvider>
        <Filter filter={filter} instances={testState.instances} />
      </RenderWithProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call dispatch with expected parameters when static filter changes', () => {
    const dispatch = jest.fn();
    const filter = testState.filtersOrdered[0];
    const { getByTestId } = render(
      <RenderWithProvider testDispatch={dispatch}>
        <Filter  filter={filter} instances={testState.instances} />
      </RenderWithProvider>
    );

    const selectFilter = getByTestId('select-Name');
    fireEvent.change(selectFilter, { target: { value: "minion" } });
    const payload = { "name": "Name", "value": "minion" };
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ type: FILTER_UPDATE, payload });
  });

  it('should call dispatch with expected parameters when dynamic filter changes', () => {
    const dispatch = jest.fn();
    const filter = testState.filtersOrdered[4];

    const expectedFilter = {
      name: 'Image Id',
      field: 'ImageId',
      type: 'dynamic',
      match: 'exact',
      values: [],
      includeEmptyDefaultValue: true
    };

    expect(filter).toMatchObject(expectedFilter);

    const { asFragment, getByTestId } = render(
      <RenderWithProvider testDispatch={dispatch}>
        <Filter  filter={filter} instances={testState.instances} />
      </RenderWithProvider>
    );

    expect(asFragment()).toMatchSnapshot();

    const selectFilter = getByTestId('select-ImageId');
    const imageId = "ami-5442621417d9c502d";
    fireEvent.change(selectFilter, { target: { value: imageId } });
    const payload = { "name": "ImageId", "value": imageId };
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ type: FILTER_UPDATE, payload });
  });
});