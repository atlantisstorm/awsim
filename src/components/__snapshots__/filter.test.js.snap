// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component should call dispatch with expected parameters when dynamic filter changes 1`] = `
<DocumentFragment>
  <div
    class="filter"
  >
    <span
      class="label"
    >
      Image Id
    </span>
    <select
      data-testid="select-ImageId"
      id="ImageId"
      name="ImageId"
    >
      <option
        value=""
      />
      <option
        value="ami-5442061867f5bff4e"
      >
        ami-5442061867f5bff4e
      </option>
      <option
        value="ami-5442621417d9c502d"
      >
        ami-5442621417d9c502d
      </option>
      <option
        value="ami-54426320e5fe48376"
      >
        ami-54426320e5fe48376
      </option>
      <option
        value="ami-54426fc4f22b6dbda"
      >
        ami-54426fc4f22b6dbda
      </option>
      <option
        value="ami-5442773270f1dbc84"
      >
        ami-5442773270f1dbc84
      </option>
      <option
        value="ami-5442a46261de55fcc"
      >
        ami-5442a46261de55fcc
      </option>
      <option
        value="ami-5442a9d2d27a71bba"
      >
        ami-5442a9d2d27a71bba
      </option>
      <option
        value="ami-5442b4753027ae935"
      >
        ami-5442b4753027ae935
      </option>
      <option
        value="ami-5442d1a288c950739"
      >
        ami-5442d1a288c950739
      </option>
      <option
        value="ami-5442e32fd1aba0456"
      >
        ami-5442e32fd1aba0456
      </option>
      <option
        value="ami-5442e32fd1aba61a6"
      >
        ami-5442e32fd1aba61a6
      </option>
      <option
        value="ami-5442e32fd1ababf88"
      >
        ami-5442e32fd1ababf88
      </option>
      <option
        value="ami-5442e32fd1abada2a"
      >
        ami-5442e32fd1abada2a
      </option>
    </select>
  </div>
</DocumentFragment>
`;

exports[`Filter component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="filter"
  >
    <span
      class="label"
    >
      Name
    </span>
    <select
      data-testid="select-Name"
      id="Name"
      name="Name"
    >
      <option
        value=""
      />
      <option
        value="ftp"
      >
        ftp
      </option>
      <option
        value="image"
      >
        image
      </option>
      <option
        value="loghost"
      >
        loghost
      </option>
      <option
        value="minion"
      >
        minion
      </option>
      <option
        value="rx"
      >
        rx
      </option>
      <option
        value="sql"
      >
        sql
      </option>
      <option
        value="tx"
      >
        tx
      </option>
      <option
        value="workhorse"
      >
        workhorse
      </option>
    </select>
  </div>
</DocumentFragment>
`;
